SWIFT-178 Add spec test runner framework

patrickfreed opened the pull request:
This is the third in a series of PRs for implementing SWIFT-178. 

This one introduces the spec test runner framework (based on Kaitlin's `crud-experiment` branch) which allows tests to be decoded from spec test JSON files. It defines structures for each of the possible CRUD operations as well as types for the tests themselves, their outcomes, and the results of operations. I refrained from introducing a "test file" construct since I figured the format would vary from test case to test case and that it would be better to leave it to the individual test runners to define them.

The retryable writes runner will use this framework in the next PR, and the crud tests should be updated to use it in later work.

kmahar commented on Tests/MongoSwiftTests/SpecTestRunner.swift:
```diff
@@ -0,0 +1,587 @@
+}
+
+/// Protocol for allowing conversion from different result types to `BulkWriteResult`.
+protocol BulkWriteResultConvertible {
```
maybe worth a comment explaining why the conversions are useful

kmahar commented on Tests/MongoSwiftTests/SpecTestRunner.swift:
```diff
@@ -0,0 +1,587 @@
+                    collection: collection,
+                    session: session)
+        } catch {
+            if case let ServerError.bulkWriteError(_, _, bulkResult, _) = error {
```
oh duh, I misread what this was doing before.

kmahar commented on Tests/MongoSwiftTests/SpecTestRunner.swift:
```diff
@@ -0,0 +1,587 @@
+
+/// Protocol for allowing conversion from different result types to `BulkWriteResult`.
+protocol BulkWriteResultConvertible {
+    var bulkResultValue: BulkWriteResult { get }
```
no strong feelings but will leave comment open in case Matt has an opinion

mbroadst replied:
I think this is fine for now

patrickfreed commented:
@mbroadst Per our offline conversation, I moved the runner stuff into its own directory ("MongoSwiftTests/SpecTestRunner", can't remember if that was the name we decided though) and split it into three files: `SpecTest.swift`, `TestOperationResult.swift`, and `TestOperation.swift`.

mbroadst commented:
@patrickfreed LGTM, let's go ahead and merge
